### PR TITLE
External query

### LINKED ISSUE #6: Use nl2graphquery for recursive retrieval

### LINKED ISSUE #7: Query external knowledge base for knowledge graph

### PR DESCRIPTION
**Pull request by kevinl424**
Recursive Retrieval

- main recursive retrieval notebooks using nl2graphquery and graph RAG 
- beginnings of llm testing, ollama vs perplexity (speed and quality)
- external query using perplexity (sonar-small model)
- custom entity extraction and synonym expansion

### CODE REVIEW

File: recursive_retrieval/docs/updates.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Recursive Retrieval Updates
+## Update 3/21/24
+- Custom synonym expansion
```
Comment (seyeong-han) [inline recursive_retrieval/docs/updates.md:3]: Impressive on improving web query using synonym expansion!

File: recursive_retrieval/entity_extraction/entity_extraction.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+from langchain_openai import OpenAI
+from langchain.prompts import PromptTemplate
+from langchain.chains import LLMChain
+from langchain_core.output_parsers import JsonOutputParser
+from typing import List
+import os
```
Comment (seyeong-han) [inline recursive_retrieval/entity_extraction/entity_extraction.py]: Usually, importing standard library should be top.
You can automatically fix those problems using `yapf` formatter!
```
pip install yapf
yapf -i [files]
```
`-i` means making changes to files in place.

File: recursive_retrieval/langchain_retrieval/retrieve.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+from langchain_openai import OpenAIEmbeddings
+from langchain_core.pydantic_v1 import BaseModel, Field
+from typing import List
+from langchain_core.prompts import ChatPromptTemplate
+from langchain_openai import ChatOpenAI
+from langchain_community.vectorstores.neo4j_vector import remove_lucene_chars
+
+
+class Entities(BaseModel):
+    """Identifying information about entities."""
```
Comment (seyeong-han) [inline recursive_retrieval/langchain_retrieval/retrieve.py:14]: Did you write all these docstring?
I love all the detailed descriptions!

**Review by seyeong-han — COMMENTED** [review]
You did a lot of experiments for a while!
Thanks for all your efforts and time :+1: 

What I understand from these External query PRs is,
- new entity_extraction module
- customizable external_perplexity module
- new synonym expand solution
- langchain retrieval test

Feel free to tell me if I missed something!

I have a question about langchain.  
Is langchain retrieval the same as llamaindex retrieval? Why did you test langchain?

**Comment by kevinl424** [comment]
> You did a lot of experiments for a while! Thanks for all your efforts and time 👍
> 
> What I understand from these External query PRs is,
> 
> * new entity_extraction module
> * customizable external_perplexity module
> * new synonym expand solution
> * langchain retrieval test
> 
> Feel free to tell me if I missed something!
> 
> I have a question about langchain. Is langchain retrieval the same as llamaindex retrieval? Why did you test langchain?

Yep, exactly. I would say the two biggest additions are the entity_extraction module, which will allow for querying the web if there is no response in the KG, and the custom synonym expander which makes querying the KG a lot more accurate. 

I tested the LangChain retrieval before I stumbled upon the custom synonym expansion solution because the llama_index retriever was as bit faulty without it. In general, the LangChain retrieval uses a similar approach (more user control over each step) but is far newer (added in last couple of weeks maybe) and thus is still a work in progress. Would be more difficult to use that approach since we are using llama_index for building the KG. 

LangChain was used however for the synonym extractor since it is very efficient for extracting structured responses in a pipeline.

File: recursive_retrieval/entity_extraction/entity_extraction.py (hunk 2 above)
Comment (kevinl424) [reply recursive_retrieval/entity_extraction/entity_extraction.py]: Fixed!

File: recursive_retrieval/langchain_retrieval/retrieve.py (hunk 3 above)
Comment (kevinl424) [reply recursive_retrieval/langchain_retrieval/retrieve.py:14]: I wrote some of them, others are from a tutorial though

**Review by seyeong-han — APPROVED** [review]
Thanks for your kind explanations!

Now, I can understand your intention of using Langchain.
Maybe we can do two tracks of llama-index and Langchain because they are revolutionizing constantly and let's keep an eye on them.

**Pull request merged**